server/src/main/java/org/opensearch/search/aggregations/bucket/terms/IncludeExclude.java

```diff
@@ -393,6 +401,91 @@ public LongBitSet acceptedGlobalOrdinals(SortedSetDocValues globalOrdinals) thro
 
     }
 
+    /**
+     * An ordinals filter that includes/excludes all ordinals corresponding to terms starting with the given prefixes
+     */
+    static class PrefixBackedOrdinalsFilter extends OrdinalsFilter {
+
+        private final SortedSet<BytesRef> includePrefixes, excludePrefixes;
+
+        private PrefixBackedOrdinalsFilter(SortedSet<BytesRef> includePrefixes, SortedSet<BytesRef> excludePrefixes) {
+            this.includePrefixes = includePrefixes;
+            this.excludePrefixes = excludePrefixes;
+        }
+
+        private static BytesRef nextBytesRef(BytesRef bytesRef) {
+            BytesRef next = BytesRef.deepCopyOf(bytesRef);
+            int pos = next.offset + next.length - 1;
+            while (pos >= next.offset && next.bytes[pos] == -1) {
+                next.bytes[pos] = 0;
+                pos--;
+            }
+            if (pos >= next.offset) {
+                next.bytes[pos]++;
+            } else {
+                // Every byte in our prefix had value 0xFF. We must match all subsequent ordinals.
+                return null;
+            }
+            return next;
+        }
+
+        private interface LongLongBiconsumer {
+            void accept(long a, long b);
+        }
+
+        private static void process(
+            SortedSetDocValues globalOrdinals,
+            long length,
+            SortedSet<BytesRef> prefixes,
+            LongLongBiconsumer consumer
+        ) throws IOException {
+            for (BytesRef prefix : prefixes) {
+                long startOrd = globalOrdinals.lookupTerm(prefix);
+                if (startOrd < 0) {
+                    // The prefix is not an exact match in the ordinals (can skip equal length below)
+                    startOrd = -1 - startOrd;
+                    // Make sure that the term at startOrd starts with prefix
+                    BytesRef startTerm = globalOrdinals.lookupOrd(startOrd);
+                    if (startTerm.length <= prefix.length
```

Comment:
`<=`? I would have assumed that if startTerm and prefix were same length we would let the consumer set the bits?

Reply:
My comment above (`can skip equal length below`) covers this but isn't very clear.

Because `startOrd` was negative, it means the prefix does not exactly match anything in the global ords. If the lengths are equal, then we know that `startTerm` cannot start with `prefix` (so we can skip the `Arrays.equals` check).

As an example, if the prefix is `aaa` and `lookupTerm` returns a negative index, then there are two possibilities:
1. `startTerm` starts with `aaa`, but has strictly more characters (e.g. `startTerm=aaaa` or `startTerm=aaaFoo`, etc.)
2. `startTerm` doesn't start with `aaa`, and can have any number of characters (e.g. `startTerm` could be `b`, `ab`, `aab`, `aaba`, etc.). Regardless, it means this prefix is not present in the index.

So, if `startTerm.length <= prefix.length`, we *must* be in the second possibility (and don't need to check the individual bytes to confirm it).